### PR TITLE
fix queue hang in split

### PR DESCRIPTION
**Pull request by filip-halt**
Also fixes ray file change.
Signed-off-by: Filip Haltmayer <filip.haltmayer@zilliz.com>

### CODE REVIEW

**Comment by reiase** [comment]
Is it possible to detect the situation when some queue is full but another has become empty? We can use non-blocking operations on the queue, so we can awake the scheduler and enlarge the queue.

**Comment by filip-halt** [comment]
> 
I think it might be. id rather put this pr through because parallel will pretty much always block without this pr. There is a timeout function on queue.get() where we could try some kind of logic. I just want to get this working so that parallel isnt broken.

File: towhee/functional/mixins/parallel.py [hunk 1]
```diff
@@ -173,9 +171,41 @@ def inner(queue):
...
+                    cach_id, cache_queue, cache_val = cached_values.pop(0)
+                    try:
+                        cache_queue.put_nowait(cache_val)
+                    except Full:
+                        cached_values.append((cach_id, cache_queue, cache_val))
+                        if first_recached is None:
+                            first_recached = cach_id
+                # Queue add logic
                 for queue in queues:
-                    queue.put(x)
```
[the remark this answers is not in the document]
Comment (filip-halt) [reply towhee/functional/mixins/parallel.py:178]: I think caching is better than increasing queue size, this allows us to cache only what we need. The cache fills with the blocked queues future inserts, allowing the other queues to catch up.

[the remark this answers is not in the document]
Comment (filip-halt) [reply towhee/functional/mixins/parallel.py:178]: image

I think this is where the hangup might be coming from. Its an executor problem.

**Comment by fzliu** [comment]
Please remember to squash these changes.

**Pull request merged**